### PR TITLE
add specialization and loop optis benchmark suites

### PR DESCRIPTION
**Pull request by tnielens**
Relates to #428.

### CODE REVIEW

File: saddle-jmh/src/main/scala/org/saddle/LoopOptimizations.scala [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+    b = scala.util.Random.nextDouble()
+  }
+
+  @Benchmark
+  /** Simplest variant, should trigger vectorization and other optimisations.
+    */
+  def array(): Array[Double] = {
+    var i = 0
+    while (i < arr.length) {
+      arr(i) = arr(i) + b
```
Comment (tnielens) [inline saddle-jmh/src/main/scala/org/saddle/LoopOptimizations.scala:33]: Interestingly, assignment operators seem to harm performance. Here `arr(i) += b` performs worse.

**Review by pityka — APPROVED** [review]
Thanks! 

Do you mind pasting in the output of jmh (the table)?

File: saddle-jmh/src/main/scala/org/saddle/Specialization.scala [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+  def vecSpecialized(): Vec[Double] = {
+    import org.saddle.ops.BinOps._
+    v1 += b
+    v1
+  }
+
+  @Benchmark
+  def vecBoxed(): Vec[Double] = {
+    import org.saddle.ops.BinOps._
+    v2 /= b
```
[the remark this answers is not in the document]
Comment (pityka) [reply saddle-jmh/src/main/scala/org/saddle/Specialization.scala:43]: here, this should be += not? (to be on pair with the other benchmarks which use +)

**Review by pityka — APPROVED** [review]

**Review by pityka — APPROVED** [review]

**Comment by tnielens** [comment]
> Thanks!
> 
> Do you mind pasting in the output of jmh (the table)?

```
[info] Benchmark                                             (size)   Mode  Cnt       Score       Error  Units
[info] LoopOptimisations.arrayAdd                             10000  thrpt   10  842083.647 ± 60290.920  ops/s
[info] LoopOptimisations.arrayAddExpandedAssignmentOperator   10000  thrpt   10  841987.825 ± 65169.011  ops/s
[info] LoopOptimisations.arrayAddHandUnrolled2                10000  thrpt   10  323300.342 ±  6380.595  ops/s
[info] LoopOptimisations.arrayAddHandUnrolled5                10000  thrpt   10  326564.688 ±  3808.226  ops/s
[info] LoopOptimisations.arrayDiv                             10000  thrpt   10  192121.527 ±  3627.980  ops/s
[info] LoopOptimisations.arrayDivHandUnrolled2                10000  thrpt   10   92827.554 ±  6555.253  ops/s
[info] LoopOptimisations.vecBinOps                            10000  thrpt   10  817415.775 ± 75246.139  ops/s
[info] LoopOptimisations.vecBinOpsMacros                      10000  thrpt   10  819505.402 ± 55060.874  ops/s
```

```
[info] Benchmark                      (size)   Mode  Cnt          Score          Error  Units
[info] Specialization.vecBoxed            10  thrpt   10   14245974.781 ±   454356.299  ops/s
[info] Specialization.vecBoxed         10000  thrpt   10      13949.961 ±     1351.311  ops/s
[info] Specialization.vecSpecialized      10  thrpt   10  109038414.822 ± 11556578.403  ops/s
[info] Specialization.vecSpecialized   10000  thrpt   10     870143.052 ±    39683.066  ops/s
```

**Pull request merged**